src/R/Core/Impl/AST/Extensions/ScopeExtensions.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Common.Core;
+using Microsoft.R.Core.AST.Arguments;
+using Microsoft.R.Core.AST.DataTypes;
+using Microsoft.R.Core.AST.Functions.Definitions;
+using Microsoft.R.Core.AST.Scopes;
+using Microsoft.R.Core.AST.Scopes.Definitions;
+using Microsoft.R.Core.AST.Statements.Definitions;
+using Microsoft.R.Core.AST.Variables;
+
+namespace Microsoft.R.Core.AST {
+    public static class ScopeExtensions {
+        /// <summary>
+        /// Enumerates scopes from the current up to the global scope.
+        /// </summary>
+        public static IEnumerable<IScope> EnumerateTowardsGlobal(this IScope scope) {
+            yield return scope;
+            while (!(scope is GlobalScope)) {
+                scope = scope.GetEnclosingScope();
+                yield return scope;
+            }
+        }
+
+        /// <summary>
+        /// Enumerates definitions of variables applicable to the given scope.
+        /// Traverses scopes from the provided scope up and enumerates
+        /// variables that appear at the left side of the left-hand assignment
+        /// operator or at the right side of the right-hand assignmend operator.
+        /// Includes both regular variables as well as function definitions.
+        /// Includes assignments that appear in the file up to the specified
+        /// position except in the global scope it enumerates all assignments.
```

Comment:
I think there should be the same behavior here regardless of the scope. In global, we want this to look ahead, because functions often call other functions (that are ahead) in their body; and body is usually evaluated after all the assignments are evaluated. But the same can be the case in non-global scopes - you can still have something like:
```r
f <- function() {
  g <- function() h()
  h <- function() ...
  g()
}
```
and it's just as valid. Similarly, something like this is not valid in global scope, but we will still incorrectly list `y` as available:
```r
x <- y
y <- 123
...
So it's a trade-off, and I think it's better if it's done the same way on both global and non-global level.